File: go/serializers/avro.go
package serializers

import (
	"fmt"
	"io"
	"reflect"
	"sync"

	"github.com/hamba/avro/v2"

	"serializer-benchmark-go/model"
	modelv2 "serializer-benchmark-go/model/v2"
)

// hambaAvro — modern, fast Apache Avro for Go (codegen-free struct binding).
// Recommended: parse schema once (cached); API.Marshal/Unmarshal for bytes;
// API.NewEncoder / NewDecoder for streams (schema-bound, Flush on Encode).
// https://github.com/hamba/avro
type hambaAvro struct {
	proto  any
	schema avro.Schema
	api    avro.API
	fxName string
}

// Package-level schema cache — Parse is not free; Prepare must not re-parse.
var (
	avroSchemaMu    sync.Mutex
	avroSchemaCache = map[string]avro.Schema{}
	avroAPI         = avro.Config{}.Freeze()
)

func newHambaAvro() *hambaAvro {
	return &hambaAvro{api: avroAPI}
}

func (s *hambaAvro) Name() string           { return "hamba/avro" }
func (s *hambaAvro) Version() string        { return ModuleVersion("github.com/hamba/avro/v2") }
func (s *hambaAvro) StreamMode() StreamMode { return StreamNative }
func (s *hambaAvro) NativeKind() NativeKind { return NativeSchema }
func (s *hambaAvro) Supports(n string) bool {
	return modelv2.IsV2TypeName(n)
}

func (s *hambaAvro) Prepare(fx model.Fixture) error {
	s.proto = fx.Value
	s.fxName = fx.Name
	sch, err := schemaFor(fx.Name)
	if err != nil {
		return err
	}
	// Batch N>1: array of records
	if isSlice(fx.Value) {
		raw := `{"type":"array","items":` + sch.String() + `}`
		sch, err = avro.Parse(raw)
		if err != nil {
			return err
		}
	}
	s.schema = sch
	return nil
}

func isSlice(v any) bool {
	if v == nil {
		return false
	}
	switch reflect.TypeOf(v).Kind() {
	case reflect.Slice, reflect.Array:
		return true
	default:
		return false
	}
}

func (s *hambaAvro) SerializeBytes(fx model.Fixture) ([]byte, error) {
	return s.api.Marshal(s.schema, fx.Value)
}

func (s *hambaAvro) DeserializeBytes(buf []byte) (any, error) {
	dst := model.NewEmptyPtr(s.proto)
	if err := s.api.Unmarshal(s.schema, buf, dst); err != nil {
		return nil, err
	}
	return model.Deref(dst), nil
}

func (s *hambaAvro) SerializeStream(fx model.Fixture, w io.Writer) (int, error) {
	cw := &countWriter{w: w}
	enc := s.api.NewEncoder(s.schema, cw)
	if err := enc.Encode(fx.Value); err != nil {
		return 0, err
	}
	return cw.n, nil
}

func (s *hambaAvro) DeserializeStream(r io.Reader) (any, error) {
	dst := model.NewEmptyPtr(s.proto)
	dec := s.api.NewDecoder(s.schema, r)
	if err := dec.Decode(dst); err != nil {
		return nil, err
	}
	return model.Deref(dst), nil
}

// Schemas use field names matching avro struct tags on data types.
// Parsed schemas are cached for the process lifetime.
func schemaFor(name string) (avro.Schema, error) {
	avroSchemaMu.Lock()
	defer avroSchemaMu.Unlock()
	if sch, ok := avroSchemaCache[name]; ok {
		return sch, nil
	}
	var raw string
	switch name {
	// Data Model v2 type_ids (JSON field names match model/v2 struct tags)
	case "message":
		raw = `{"type":"record","name":"Message","fields":[
			{"name":"f_bool","type":"boolean"},{"name":"f_int32","type":"int"},
			{"name":"f_int64","type":"long"},{"name":"f_float64","type":"double"},
			{"name":"f_string","type":"string"},{"name":"f_bool_2","type":"boolean"},
			{"name":"f_int32_2","type":"int"},{"name":"f_string_2","type":"string"}
		]}`
	case "document":
		raw = `{"type":"record","name":"Document","fields":[
			{"name":"id","type":"string"},{"name":"status","type":"int"},
			{"name":"meta","type":{"type":"record","name":"DocumentMeta","fields":[
				{"name":"region","type":"string"},{"name":"version","type":"int"}
			]}},
			{"name":"items","type":{"type":"array","items":{"type":"record","name":"DocumentItem","fields":[
				{"name":"sku","type":"string"},{"name":"qty","type":"int"},{"name":"price_minor","type":"long"}
			]}}}
		]}`
	case "telemetry":
		raw = `{"type":"record","name":"TelemetryV2","fields":[
			{"name":"source","type":"string"},{"name":"ts","type":"long"},
			{"name":"tags","type":{"type":"array","items":"string"}},
			{"name":"values","type":{"type":"array","items":"double"}}
		]}`
	case "strings":
		raw = `{"type":"record","name":"Strings","fields":[
			{"name":"items","type":{"type":"array","items":"string"}}
		]}`
	case "event":
		raw = `{"type":"record","name":"Event","fields":[
			{"name":"event_id","type":"string"},{"name":"event_type","type":"string"},
			{"name":"occurred_at","type":"long"},{"name":"producer","type":"string"},
			{"name":"attrs","type":{"type":"array","items":{"type":"record","name":"EventAttr","fields":[
				{"name":"key","type":"string"},{"name":"value","type":"string"}
			]}}}
		]}`
	default:
		return nil, fmt.Errorf("avro: no schema for %s", name)
	}
	sch, err := avro.Parse(raw)
	if err != nil {
		return nil, err
	}
	avroSchemaCache[name] = sch
	return sch, nil
}

File: swift/Sources/SerializerBenchmarkCore/Serializers/Avro.swift
import Foundation
import SwiftAvroCore

// MARK: - Apache Avro (SwiftAvroCore)
// Optimal: Avro().encodeFrom / decodeFrom with schema prepared outside the timed loop.
// Schemas match suite field names (f_bool, …). Wrappers select schema by fixture.name only.
// https://github.com/lynixliu/SwiftAvroCore

public final class SwiftAvroSerializer: BenchSerializer {
    public let name = "SwiftAvroCore"
    public let version: String
    public let streamMode: StreamMode = .adapted
    public let nativeKind: NativeKind = .schema

    private let avro = Avro()
    private var prepared: Fixture?
    private var schema: AvroSchema?
    /// Bound in prepare so timed path has no multi-way fixture switch (issue #59).
    private var encodeFn: ((Fixture) throws -> Data)?
    private var decodeFn: ((Data) throws -> Any)?

    public init() {
        self.version = PackageVersions.version(for: "swiftavrocore", fallback: "2.x")
    }

    public func supports(testDataName: String) -> Bool { true }

    public func prepare(_ fixture: Fixture) throws {
        prepared = fixture
        let schemaJSON = AvroSchemas.json(for: fixture)
        guard let sch = avro.decodeSchema(schema: schemaJSON) else {
            throw BenchError.unsupported("avro schema parse failed for \(fixture.name)")
        }
        schema = sch
        // Monomorphic encode/decode bound outside the timer (issue #59).
        encodeFn = AvroSchemas.bindEncode(avro: avro, schema: sch, fixture: fixture)
        decodeFn = AvroSchemas.bindDecode(avro: avro, schema: sch, fixture: fixture)
        // warm
        _ = try serializeBytes(fixture)
    }

    public func serializeBytes(_ fixture: Fixture) throws -> Data {
        guard let encodeFn else { throw BenchError.prepareRequired }
        return try encodeFn(fixture)
    }

    public func deserializeBytes(_ data: Data) throws -> Any {
        guard let decodeFn else { throw BenchError.prepareRequired }
        return try decodeFn(data)
    }
}

/// Avro JSON schemas + encode/decode helpers (domain-aware, not in timed wrapper surface).
enum AvroSchemas {
    static func json(for fixture: Fixture) -> String {
        if fixture.instanceCount > 1 {
            return arraySchema(item: recordSchema(typeId: fixture.name))
        }
        return recordSchema(typeId: fixture.name)
    }

    private static func arraySchema(item: String) -> String {
        """
        {"type":"array","items":\(item)}
        """
    }

    private static func recordSchema(typeId: String) -> String {
        switch typeId {
        case "message":
            return """
            {"type":"record","name":"Message","fields":[
              {"name":"f_bool","type":"boolean"},
              {"name":"f_int32","type":"int"},
              {"name":"f_int64","type":"long"},
              {"name":"f_float64","type":"double"},
              {"name":"f_string","type":"string"},
              {"name":"f_bool_2","type":"boolean"},
              {"name":"f_int32_2","type":"int"},
              {"name":"f_string_2","type":"string"}
            ]}
            """
        case "document":
            return """
            {"type":"record","name":"Document","fields":[
              {"name":"id","type":"string"},
              {"name":"status","type":"int"},
              {"name":"meta","type":{"type":"record","name":"DocumentMeta","fields":[
                {"name":"region","type":"string"},{"name":"version","type":"int"}
              ]}},
              {"name":"items","type":{"type":"array","items":{"type":"record","name":"DocumentItem","fields":[
                {"name":"sku","type":"string"},{"name":"qty","type":"int"},{"name":"price_minor","type":"long"}
              ]}}}
            ]}
            """
        case "telemetry":
            return """
            {"type":"record","name":"Telemetry","fields":[
              {"name":"source","type":"string"},
              {"name":"ts","type":"long"},
              {"name":"tags","type":{"type":"array","items":"string"}},
              {"name":"values","type":{"type":"array","items":"double"}}
            ]}
            """
        case "strings":
            return """
            {"type":"record","name":"Strings","fields":[
              {"name":"items","type":{"type":"array","items":"string"}}
            ]}
            """
        case "event":
            return """
            {"type":"record","name":"Event","fields":[
              {"name":"event_id","type":"string"},
              {"name":"event_type","type":"string"},
              {"name":"occurred_at","type":"long"},
              {"name":"producer","type":"string"},
              {"name":"attrs","type":{"type":"array","items":{"type":"record","name":"EventAttr","fields":[
                {"name":"key","type":"string"},{"name":"value","type":"string"}
              ]}}}
            ]}
            """
        default:
            return """
            {"type":"record","name":"Unknown","fields":[]}
            """
        }
    }

    /// Select monomorphic encode once in prepare (no timed multi-way switch).
    static func bindEncode(avro: Avro, schema: AvroSchema, fixture: Fixture) -> (Fixture) throws -> Data {
        let batch = fixture.instanceCount > 1
        switch fixture.name {
        case "message":
            return batch
                ? { fx in try avro.encodeFrom(fx.value as! [Message], schema: schema) }
                : { fx in try avro.encodeFrom(fx.value as! Message, schema: schema) }
        case "document":
            return batch
                ? { fx in try avro.encodeFrom(fx.value as! [Document], schema: schema) }
                : { fx in try avro.encodeFrom(fx.value as! Document, schema: schema) }
        case "telemetry":
            return batch
                ? { fx in try avro.encodeFrom(fx.value as! [Telemetry], schema: schema) }
                : { fx in try avro.encodeFrom(fx.value as! Telemetry, schema: schema) }
        case "strings":
            return batch
                ? { fx in try avro.encodeFrom(fx.value as! [Strings], schema: schema) }
                : { fx in try avro.encodeFrom(fx.value as! Strings, schema: schema) }
        case "event":
            return batch
                ? { fx in try avro.encodeFrom(fx.value as! [Event], schema: schema) }
                : { fx in try avro.encodeFrom(fx.value as! Event, schema: schema) }
        default:
            let name = fixture.name
            return { _ in throw BenchError.unknownType(name) }
        }
    }

    static func bindDecode(avro: Avro, schema: AvroSchema, fixture: Fixture) -> (Data) throws -> Any {
        let batch = fixture.instanceCount > 1
        switch fixture.name {
        case "message":
            return batch
                ? { data in try avro.decodeFrom(from: data, schema: schema) as [Message] }
                : { data in try avro.decodeFrom(from: data, schema: schema) as Message }
        case "document":
            return batch
                ? { data in try avro.decodeFrom(from: data, schema: schema) as [Document] }
                : { data in try avro.decodeFrom(from: data, schema: schema) as Document }
        case "telemetry":
            return batch
                ? { data in try avro.decodeFrom(from: data, schema: schema) as [Telemetry] }
                : { data in try avro.decodeFrom(from: data, schema: schema) as Telemetry }
        case "strings":
            return batch
                ? { data in try avro.decodeFrom(from: data, schema: schema) as [Strings] }
                : { data in try avro.decodeFrom(from: data, schema: schema) as Strings }
        case "event":
            return batch
                ? { data in try avro.decodeFrom(from: data, schema: schema) as [Event] }
                : { data in try avro.decodeFrom(from: data, schema: schema) as Event }
        default:
            let name = fixture.name
            return { _ in throw BenchError.unknownType(name) }
        }
    }
}
